llvm/include/llvm/BinaryFormat/DynamicTags.def

```diff
@@ -86,6 +86,8 @@ DYNAMIC_TAG(RELRSZ, 35)  // Size of Relr relocation table.
 DYNAMIC_TAG(RELR, 36)    // Address of relocation table (Relr entries).
 DYNAMIC_TAG(RELRENT, 37) // Size of a Relr relocation entry.
 
+DYNAMIC_TAG(CREL,  38)   // CREL relocation table
+
```

Comment:
I think we need to avoid generic numbers for dynamic tags/section types etc, except possibly generic numbers that are a long way from the "normal" range. The reason for this is we don't know what values are going to get standardised, assuming this feature even gets accepted into the standard. If, for example, another feature gets accepted into the standard before CREL, it might use up the slot you've "allocated" for the DT_CREL here, which could cause problems with loaders that have supported the experimental CREL implementation. On the other hand, a different special number (whether in a dedicated range or otherwise) that isn't adjacent to the list will not have the issue of a potential clash.

Reply:
> it might use up the slot you've "allocated"

I understand the concern but this is just impossible given the essential state of the generic ABI...

(I have some notes at https://maskray.me/blog/2024-05-26-evolution-of-elf-object-file-format#a-future-in-flux)

Reply:
> I understand the concern but this is just impossible given the essential state of the generic ABI...

I don't understand this comment. There's an active mailing list, where proposals are discussed and adopted, even if there's no fully published document, so a different proposal could easily come up whilst this is still under review. Perhaps it'll be implemented in GNU/Solaris/... initially with the same generic value. Whose meaning of the value should then win?

In the worst case, maybe the generic ABI list ends up rejecting this crel proposal. The values aren't reserved for LLVM usage, so the next proposal to come along will use those same values and we then have a clash between what LLVM wants to use the value for (and there might be objects out there using that value, if this feature is adopted in an LLVM release), which would cause us all sorts of headaches in the future.

Reply:
When I have more bandwidth to implement the DT_CREL support for binutils and some loader implementations, I'll use 38.
This avoids the potential annoyance of picking a version initially and then needing to change it later, which could cause issues with toolchain/loader interop inconvenience.

In practice, achieving consensus between major ELF toolchain vendors like GNU and LLVM is enough to prevent conflicts...
The generic ABI hasn't seen a new `DT_*` addition besides RELR since 2013.

Reply:
> I think we need to avoid generic numbers for dynamic tags/section types etc, except possibly generic numbers that are a long way from the "normal" range. 

+1.

I've tried to pushback on using any standard range before it's been standardized (eg: https://github.com/llvm/llvm-project/pull/91280#issuecomment-2099201183 https://github.com/llvm/llvm-project/pull/91280#issuecomment-2101019352 ) but I haven't looked through the change completely to see all the enumerations used and how they're reserved, whether they use an extension space or what the argument is for them not using one.

Perhaps someone could summarize this aspect of the patch?

Reply:
I changed DT_CREL to 0x40000026 for now.

It's complex but the spec on sco.com might get further updates, but it doesn't matter if GNU and LLVM can be kept in sync and the specification is essentially in the public domain. https://news.ycombinator.com/item?id=40486068